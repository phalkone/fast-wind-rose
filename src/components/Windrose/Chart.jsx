/* eslint-disable react/jsx-fragments */
import React, { Fragment } from 'react'
import PropTypes from 'prop-types'

/**
 * The chart that can be divided into a specified number of sectors
 */
function Chart (props) {
  return (
    <Fragment>
      {/* Sector lines */}
      {[...new Array(props.sectorCount).keys()].map(i => (
        <Fragment key={i}>
          <line
            x1={props.center}
            x2={props.center}
            y1={10}
            y2={props.center}
            stroke='grey'
            strokeWidth='1'
            transform={`rotate(${(0.5 + i) * (360 / props.sectorCount)},` +
                       ` ${props.center},${props.center})`}
          />
        </Fragment>
      ))}
      {/* Circles */}
      {[...new Array(4).keys()].map(i => (
        <Fragment key={i}>
          <circle
            cx={props.center}
            cy={props.center}
            r={(4 - i) * ((props.center - 10) / 4)}
            fill='none'
            strokeWidth='1'
            stroke='grey'
            opacity={i > 0 ? 0.5 : 1}
          />
        </Fragment>
      ))}
    </Fragment>
  )
}

Chart.propTypes = {
  /**
   * The number of sectors
   */
  sectorCount: PropTypes.number,
  /**
   * The center of the chart. x and y coordinates are the same.
   */
  center: PropTypes.number
}

export default Chart
