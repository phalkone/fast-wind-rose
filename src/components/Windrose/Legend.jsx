/* eslint-disable react/jsx-fragments */
import React, { Fragment } from 'react'
import PropTypes from 'prop-types'

/**
 * The legend displaying the used scale
 */
function Legend (props) {
  const keys = Object.keys(props.scale)
  const length = keys.length
  const square = ((props.size - 20) / length) > 20 ? 20 : ((props.size - 20) / length)
  const margin = (props.size - (length * square)) / 2
  const fontSize = square - 2

  return (
    <Fragment>
      {keys.map((speed, i) => (
        <Fragment key={speed}>
          {/* Color scale */}
          <rect
            x={props.size + 10}
            y={(props.size - margin) - ((i + 1) * square)}
            width={square}
            height={square}
            stroke='grey'
            strokeWidth='1'
            fill={props.scale[speed]}
          />
          {/* Speed scale */}
          <text
            x={props.size + square + 15}
            y={(props.size - margin) - (i * square) + 5}
            fill='black'
            fontSize={fontSize}
          >{i === length - 1 ? `${speed}<` : speed}
          </text>
        </Fragment>
      ))}
      {/* Unit */}
      <text
        x={props.size + square + 15}
        y={(props.size - margin) - (length * square) + 5}
        fill='black'
        fontSize={fontSize}
      >kts
      </text>
    </Fragment>
  )
}

Legend.propTypes = {
  /**
   * Size of chart (height/width + margins)
   */
  size: PropTypes.number,
  /**
   * Scale of speeds with the linked color. Example as follows:
   *  { 0: 'rgb(60,95,156)', 5: 'rgb(94,131,188)' }
   */
  scale: PropTypes.object
}

export default Legend
