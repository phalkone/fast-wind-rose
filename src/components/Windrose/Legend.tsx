import React, { Fragment, useContext } from 'react'
import { WindroseContext } from '.'

/**
 * The legend displaying the used scale
 */
export const Legend = () => {
  const context = useContext(WindroseContext)
  const size = context.center * 2
  const keys = Object.keys(context.scale)
  const length = keys.length
  const square = ((size - 20) / length) > 20 ? 20 : ((size - 20) / length)
  const margin = (size - (length * square)) / 2
  const fontSize = square - 2

  return (
    <>
      {keys.map((speed, i) => (
        <Fragment key={speed}>
          {/* Color scale */}
          <rect
            x={size + 10}
            y={(size - margin) - ((i + 1) * square)}
            width={square}
            height={square}
            stroke='grey'
            strokeWidth='1'
            fill={context.scale[speed]}
          />
          {/* Speed scale */}
          <text
            x={size + square + 15}
            y={(size - margin) - (i * square) + 5}
            fill='black'
            fontSize={fontSize}
          >{i === length - 1 ? `${speed}<` : speed}
          </text>
        </Fragment>
      ))}
      {/* Unit */}
      <text
        x={size + square + 15}
        y={(size - margin) - (length * square) + 5}
        fill='black'
        fontSize={fontSize}
      >kts
      </text>
    </>
  )
}
