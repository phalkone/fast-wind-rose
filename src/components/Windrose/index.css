button {
  position: absolute;
  top: 10px;
  color: rgb(0, 0, 0);
  background-color: white;
  font-weight: bolder;
  border: 1px solid black;
  text-decoration: none;
  padding: 0 5px;
  transition-duration: 0.4s;
}

button:hover {
  background-color: rgb(203, 203, 203);
}

select {
  position: absolute;
  top: 10px;
  left: 10px;
  padding: 2px;
}

svg, select, button {
  font-family: Roboto, "Open Sans", sans-serif;
}
