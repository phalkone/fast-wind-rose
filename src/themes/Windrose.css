.windrose-button{
  position: absolute;
  top: 10px;
  color: rgb(0, 0, 0);
  background-color: white;
  font-weight: bolder;
  border: 1px solid black;
  text-decoration: none;
  padding: 0 5px;
  transition-duration: 0.4s;
}
 
.windrose-button:hover {
  background-color: rgb(203, 203, 203);
}

.windrose-select {
  position: absolute;
  top: 10px;
  left: 10px;
  padding: 2px;
}

.windrose-container {
  font-family: Roboto, "Open Sans", sans-serif;
}