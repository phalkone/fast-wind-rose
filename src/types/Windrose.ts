export interface IWindrose {
  /**
   * Hide or display legend
   */
  legend: boolean,
  /**
   * Width/height of chart. Will be displayed in specified viewbox.
   */
  size: number,
  /**
   * Default number of sectors. Must be included in sectorArray.
   */
  sectorCount: number,
  /**
   * Fixed interval between data points specified in hours. For example if
   * data points are every 30 min, then value should be 0.5
   */
  interval: number,
  /**
   * Scale of speeds with the linked color. Example as follows:
   *  { 0: 'rgb(60,95,156)', 5: 'rgb(94,131,188)' }
   */
  scale: { [n: number]: string },
  /**
   * Array with directional data for the wind
   */
  dirData: object[],
  /**
  * Array with speed data for the wind
  */
  spdData: object[],
  /**
   * Key for speed data
   */
  spdKey: string,
  /**
   * Key for directional data
   */
  dirKey: string,
  /**
   * Scales the compilation scale against the viewbox. Choose a value lower
   * than 1 to scale down and larger than 1 to scale up.
   */
  enlarge: number,
  /**
   * Array that defines the number of sectors the user can choose from.
   */
  sectorArray: number[],
  /**
   * Common key between direction and speed data
   */
  commonKey: string
}

export interface ILegend {
  /**
   * Size of the chart. Width = Height.
   */
  size: number,
  /**
   * Scale of speeds with the linked color. Example as follows:
   *  { 0: 'rgb(60,95,156)', 5: 'rgb(94,131,188)' }
   */
  scale: { [n: number]: string }
}

export interface IShip {
  /**
   * The center of the ship path. x and y coordinates are assumed to be the same.
   */
  center: number
}

export interface IChart extends IShip {
  /**
   * The number of sectors
   */
  sectorSize: number
}

export interface IIntervalLabel extends IChart {
  /**
   * Defines which sector is being drawn. 0 being the first sector.
   */
  sector: number,
  /**
   * Interval between data points specified in hours
   */
  interval: number,
  /**
   * Array with the speeds that are part of this sector
   */
  speeds: number[]
}

export interface ISector extends IIntervalLabel, ILegend {
  /**
   * Total bar length of this sector. The sector with the largest interval will
   * have a bar length equal to the radius.
   */
  barLength: number,
  /**
   * Defines the bar length of 1 unit of interval.
   */
  unit: number,
}

export interface IToolTip {
  /**
   * x coordinate of top left corner of tooltip
   */
  x: number,
  /**
   * y coordinate of top left corner of tooltip
   */
  y: number,
  /**
   * text of tooltip
   */
  text: [string, number],
  /**
   * Scale factor due to viewBox of SVG vs actual size of SVG.
   */
  factor: number
}
