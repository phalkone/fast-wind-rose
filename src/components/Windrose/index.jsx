import React, { Fragment } from 'react'
import Legend from './Legend'
import Chart from './Chart'
import Ship from './Ship'
import Sector from './Sector'
import IntervalLabel from './IntervalLabel'
import PropTypes from 'prop-types'

export default class Windrose extends React.Component {
  constructor (props) {
    super(props)
    const sectors = [...new Array(360 / this.props.sectorSize).keys()].map((el) => [])
    for (let i = 0; i < this.props.data.length; i++) {
      let dir = this.props.data[i].direction + (this.props.sectorSize / 2)
      if (dir > 360) dir -= 360
      const cat = Math.floor(dir / this.props.sectorSize)
      sectors[cat].push(this.props.data[i].speed)
    }
    this.max = sectors.map((el) => el.length).sort((a, b) => b - a)[0]
    this.sectors = sectors
  }

  renderSectors () {
    return (
      this.sectors.map((speeds, i) => (
        <Fragment key={i}>
          {speeds.length &&
            <IntervalLabel
              sector={i}
              interval={speeds.length}
              sectorSize={this.props.sectorSize}
            />}
          {speeds.length &&
            <Sector
              sector={i}
              speeds={speeds}
              sectorSize={this.props.sectorSize}
              barLength={(speeds.length / this.max) * 120}
              interval={120 / this.max}
            />}
        </Fragment>
      ))
    )
  }

  render () {
    return (
      <svg
        version='1.1'
        viewBox='0 0 325 260'
        width={this.props.width}
        height={this.props.height}
      >
        {this.props.legend && <Legend />}
        <Chart sectorSize={this.props.sectorSize} />
        {this.renderSectors()}
        <Ship />
      </svg>
    )
  }
}

Windrose.propTypes = {
  height: PropTypes.number,
  width: PropTypes.number,
  sectorSize: PropTypes.number,
  legend: PropTypes.bool,
  data: PropTypes.array
}
