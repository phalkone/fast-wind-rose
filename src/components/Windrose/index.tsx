import React, { createContext, Fragment, useState } from 'react'
import { Legend } from './Legend'
import { Chart } from './Chart'
import { Ship } from './Ship'
import { Sector } from './Sector'
import { divideBySector } from '../../utils/Windrose.util'
import type { IWindrose, IWindroseContext } from '../../types/Windrose.types'
import '../../themes/Windrose.scss'

/**
 * Draws a windrose for provided relative wind data
 * @param {IWindrose} props Properties of the Windrose as descibed by IWindrose.
 */
const Windrose = (props: IWindrose) => {
  /* Use state to set the number of sectors and show/hide legend */
  const [sectorCount, setSectorCount] = useState<number>(props.sectorCount)
  const [legend, setLegend] = useState<boolean>(props.legend)

  /* Divide data points per sector */
  const sectors = divideBySector(sectorCount, props.dirData, props.dirKey,
    props.spdData, props.spdKey, props.commonKey)

  /* Set the viewBox size compared to the compilation size */
  const compilationSize = 260 / props.enlarge

  /* Add sectorCount if not part of sectorArray */
  if (!props.sectorArray.includes(sectorCount)) {
    props.sectorArray.push(sectorCount)
  }

  return (
    <div className='windrose-container'>
      {/* Selector for the number of sectors */}
      <select
        title='Number of sectors'
        value={sectorCount}
        className='windrose-select'
        onChange={e => { setSectorCount(Number(e.target.value)) }}
      >
        {props.sectorArray.map(s => (
          <Fragment key={s}>
            <option value={s}>{s}</option>
          </Fragment>
        ))}
      </select>
      <button
        title='Show/hide legend'
        className='windrose-button'
        onClick={e => { setLegend(!legend) }}
        style={{
          left: props.size - 30
        }}
      >{legend ? '<<' : '>>'}
      </button>
      {/* Windrose SVG with optional legend */}
      <svg
        version='1.1'
        viewBox={`0 0 ${compilationSize + (legend ? 65 : 0)} ${compilationSize}`}
        width={props.size + (legend ? 65 * (props.size / compilationSize) : 0)}
        height={props.size}
      >
        <WindroseContext.Provider value={{
          center: compilationSize / 2,
          scale: props.scale,
          sectorSize: 360 / sectorCount,
          interval: props.interval,
          size: props.size,
          max: sectors.map((el) => el.length).sort((a, b) => b - a)[0]
        }}>
          {/* Draw legend only if option is given as prop */}
          {legend && <Legend/>}
          <Chart/>
          {/* Draw each sector with interval label */}
          {sectors.map((speeds, i) => (
            speeds.length && <Fragment key={i}>
              <Sector
                sector={i}
                speeds={speeds}
              />
            </Fragment>
          ))}
          {/* Draw ship outline */}
          <Ship/>
        </WindroseContext.Provider>
        {/* Ensures that tooltips are always on top */}
        <use href='#tooltip' fill='black' />
        <use href='#tooltiptext' />
      </svg>
    </div>
  )
}

Windrose.defaultProps = {
  size: 260,
  spdKey: 'value',
  dirKey: 'value',
  commonKey: 'timestamp',
  sectorCount: 12,
  enlarge: 1,
  interval: 1,
  sectorArray: [4, 8, 12, 16, 24, 32, 36],
  scale: {
    0: 'rgb(60,95,156)',
    5: 'rgb(94,131,188)',
    10: 'rgb(143,180,232)',
    15: 'rgb(174,203,214)',
    20: 'rgb(220,226,220)',
    25: 'rgb(251,233,94)',
    30: 'rgb(252,195,67)',
    35: 'rgb(245,136,42)',
    40: 'rgb(242,103,33)',
    45: 'rgb(243,63,29)',
    50: 'rgb(244,36,27)'
  }
}

const defaultContext : IWindroseContext = {
  center: 130,
  scale: Windrose.defaultProps.scale,
  sectorSize: 30,
  interval: 1,
  size: 260,
  max: 24
}

export const WindroseContext = createContext(defaultContext)

export default Windrose
