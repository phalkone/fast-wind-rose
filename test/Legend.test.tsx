import React from 'react'
import { render, unmountComponentAtNode } from 'react-dom'
import { act } from 'react-dom/test-utils'
import { Legend } from '../src/components/Windrose/Legend'
import { WindroseContext } from '../src/components/Windrose'
import type { IWindroseContext } from '../src/types/Windrose.types'
import renderer from 'react-test-renderer'

let container : HTMLElement
beforeEach(() => {
  container = document.createElement('svg')
  container.setAttribute('viewBox', '0 0 325 260')
  document.body.appendChild(container)
})

afterEach(() => {
  unmountComponentAtNode(container)
  container.remove()
})

const context : IWindroseContext = {
  center: 130,
  scale: {
    0: 'rgb(60,95,156)',
    5: 'rgb(94,131,188)',
    10: 'rgb(143,180,232)',
    15: 'rgb(174,203,214)',
    20: 'rgb(220,226,220)',
    25: 'rgb(251,233,94)',
    30: 'rgb(252,195,67)',
    35: 'rgb(245,136,42)',
    40: 'rgb(242,103,33)',
    45: 'rgb(243,63,29)',
    50: 'rgb(244,36,27)'
  },
  sectorSize: 30,
  interval: 1,
  size: 260,
  max: 24
}

it('renders the correct number of text elements', () => {
  act(() => {
    render(
      <WindroseContext.Provider value={ context }>
        <Legend/>
      </WindroseContext.Provider>
      , container)
  })
  expect(Array.from(container.children).filter((el : Element) => el.tagName === 'TEXT').length).toEqual(12)
})

it('renders the correct number of rectangle elements', () => {
  act(() => {
    render(
      <WindroseContext.Provider value={ context }>
        <Legend/>
      </WindroseContext.Provider>
      , container)
  })
  expect(Array.from(container.children).filter((el : Element) => el.tagName === 'RECT').length).toEqual(11)
})

it('fills the colors of the rectangles as per scale', () => {
  act(() => {
    render(
      <WindroseContext.Provider value={ context }>
        <Legend/>
      </WindroseContext.Provider>
      , container)
  })
  const colors = Array.from(container.children)
    .filter((el : Element) => el.tagName === 'RECT')
    .map((el: Element) => el.getAttribute('fill'))
  expect(colors).toEqual(Object.values(context.scale))
})

it('renders the text as per scale', () => {
  act(() => {
    render(
      <WindroseContext.Provider value={ context }>
        <Legend/>
      </WindroseContext.Provider>
      , container)
  })
  const text = Array.from(container.children)
    .filter((el : Element) => el.tagName === 'TEXT')
    .map((el: Element) => el.innerHTML)

  const contextScale = Object.keys(context.scale)
  contextScale[contextScale.length - 1] = contextScale[contextScale.length - 1] + '&lt;'
  contextScale.push('kts')
  expect(text).toEqual(contextScale)
})

test('Render a Legend', () => {
  const comp = renderer.create(
    <WindroseContext.Provider value={ context }>
      <Legend/>
    </WindroseContext.Provider>
  )
  const tree = comp.toJSON()
  expect(tree).toMatchSnapshot()
})
