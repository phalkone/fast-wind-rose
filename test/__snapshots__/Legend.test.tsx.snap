// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render a Legend 1`] = `
Array [
  <rect
    fill="rgb(60,95,156)"
    height={20}
    stroke="grey"
    strokeWidth="1"
    width={20}
    x={270}
    y={220}
  />,
  <text
    fill="black"
    fontSize={18}
    x={295}
    y={245}
  >
    0
  </text>,
  <rect
    fill="rgb(94,131,188)"
    height={20}
    stroke="grey"
    strokeWidth="1"
    width={20}
    x={270}
    y={200}
  />,
  <text
    fill="black"
    fontSize={18}
    x={295}
    y={225}
  >
    5
  </text>,
  <rect
    fill="rgb(143,180,232)"
    height={20}
    stroke="grey"
    strokeWidth="1"
    width={20}
    x={270}
    y={180}
  />,
  <text
    fill="black"
    fontSize={18}
    x={295}
    y={205}
  >
    10
  </text>,
  <rect
    fill="rgb(174,203,214)"
    height={20}
    stroke="grey"
    strokeWidth="1"
    width={20}
    x={270}
    y={160}
  />,
  <text
    fill="black"
    fontSize={18}
    x={295}
    y={185}
  >
    15
  </text>,
  <rect
    fill="rgb(220,226,220)"
    height={20}
    stroke="grey"
    strokeWidth="1"
    width={20}
    x={270}
    y={140}
  />,
  <text
    fill="black"
    fontSize={18}
    x={295}
    y={165}
  >
    20
  </text>,
  <rect
    fill="rgb(251,233,94)"
    height={20}
    stroke="grey"
    strokeWidth="1"
    width={20}
    x={270}
    y={120}
  />,
  <text
    fill="black"
    fontSize={18}
    x={295}
    y={145}
  >
    25
  </text>,
  <rect
    fill="rgb(252,195,67)"
    height={20}
    stroke="grey"
    strokeWidth="1"
    width={20}
    x={270}
    y={100}
  />,
  <text
    fill="black"
    fontSize={18}
    x={295}
    y={125}
  >
    30
  </text>,
  <rect
    fill="rgb(245,136,42)"
    height={20}
    stroke="grey"
    strokeWidth="1"
    width={20}
    x={270}
    y={80}
  />,
  <text
    fill="black"
    fontSize={18}
    x={295}
    y={105}
  >
    35
  </text>,
  <rect
    fill="rgb(242,103,33)"
    height={20}
    stroke="grey"
    strokeWidth="1"
    width={20}
    x={270}
    y={60}
  />,
  <text
    fill="black"
    fontSize={18}
    x={295}
    y={85}
  >
    40
  </text>,
  <rect
    fill="rgb(243,63,29)"
    height={20}
    stroke="grey"
    strokeWidth="1"
    width={20}
    x={270}
    y={40}
  />,
  <text
    fill="black"
    fontSize={18}
    x={295}
    y={65}
  >
    45
  </text>,
  <rect
    fill="rgb(244,36,27)"
    height={20}
    stroke="grey"
    strokeWidth="1"
    width={20}
    x={270}
    y={20}
  />,
  <text
    fill="black"
    fontSize={18}
    x={295}
    y={45}
  >
    50&lt;
  </text>,
  <text
    fill="black"
    fontSize={18}
    x={295}
    y={25}
  >
    kts
  </text>,
]
`;
